File: frontend/lib/core/network/api_client.dart
import 'dart:convert';

import 'package:dio/dio.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

import '../constants.dart';

class ApiClient {
  ApiClient._();

  static final Dio _dio = Dio(BaseOptions(
    baseUrl: AppConstants.baseUrl,
    connectTimeout: const Duration(seconds: 15),
    receiveTimeout: const Duration(seconds: 15),
    headers: {'Content-Type': 'application/json'},
  ));

  static const _storage = FlutterSecureStorage();
  static bool _initialized = false;

  /// Current tenant schema sent with every request.
  static String? _tenantSchema;

  static Dio get instance {
    if (!_initialized) {
      _dio.interceptors.add(_AuthInterceptor());
      _initialized = true;
    }
    return _dio;
  }

  /// Set the active tenant schema (e.g. 'hospital_demo', 'pharmacy_demo').
  static Future<void> setTenantSchema(String? schema) async {
    _tenantSchema = schema;
    if (schema != null) {
      await _storage.write(key: AppConstants.tenantSchemaKey, value: schema);
    } else {
      await _storage.delete(key: AppConstants.tenantSchemaKey);
    }
  }

  /// Get the current tenant schema.
  static String? get tenantSchema => _tenantSchema;

  /// Restore tenant schema from storage (call at startup).
  static Future<void> restoreTenantSchema() async {
    _tenantSchema = await _storage.read(key: AppConstants.tenantSchemaKey);
  }

  static Future<void> setTokens({
    required String access,
    required String refresh,
  }) async {
    await _storage.write(key: AppConstants.accessTokenKey, value: access);
    await _storage.write(key: AppConstants.refreshTokenKey, value: refresh);
  }

  static Future<void> clearTokens() async {
    await _storage.delete(key: AppConstants.accessTokenKey);
    await _storage.delete(key: AppConstants.refreshTokenKey);
    await _storage.delete(key: AppConstants.userKey);
  }

  static Future<String?> get accessToken =>
      _storage.read(key: AppConstants.accessTokenKey);

  static Future<String?> get refreshToken =>
      _storage.read(key: AppConstants.refreshTokenKey);

  static Future<void> saveUser(Map<String, dynamic> user) async {
    await _storage.write(key: AppConstants.userKey, value: jsonEncode(user));
  }

  static Future<Map<String, dynamic>?> getUser() async {
    final data = await _storage.read(key: AppConstants.userKey);
    if (data == null) return null;
    return jsonDecode(data) as Map<String, dynamic>;
  }
}

class _AuthInterceptor extends Interceptor {
  @override
  void onRequest(
      RequestOptions options, RequestInterceptorHandler handler) async {
    final token = await ApiClient.accessToken;
    if (token != null) {
      options.headers['Authorization'] = 'Bearer $token';
    }
    // Attach tenant schema header (skip for public endpoints)
    final path = options.path;
    final isPublicEndpoint = path.contains('/tenants/register') ||
        path.contains('/auth/login') ||
        path.contains('/auth/register/');
    if (!isPublicEndpoint) {
      final schema = ApiClient.tenantSchema;
      if (schema != null) {
        options.headers['X-Tenant-Schema'] = schema;
      }
    }
    handler.next(options);
  }

  @override
  void onError(DioException err, ErrorInterceptorHandler handler) async {
    // Auto token-refresh is temporarily disabled.
    // 401s are passed through so screens can show their existing data
    // rather than silently wiping tokens and causing empty states.
    handler.next(err);
  }
}

File: nuxtfrontend/plugins/api.client.js
// Axios client - mirrors lib/core/network/api_client.dart
import axios from 'axios'
import { AppConstants } from '~/utils/constants'

const PUBLIC_ENDPOINTS = [
  '/tenants/register',
  '/auth/login',
  '/auth/register/'
]

function isPublicEndpoint(url = '') {
  return PUBLIC_ENDPOINTS.some((p) => url.includes(p))
}

export default defineNuxtPlugin((nuxtApp) => {
  const config = useRuntimeConfig()

  const api = axios.create({
    baseURL: config.public.apiBase,
    timeout: 30000,
    headers: { 'Content-Type': 'application/json' }
  })

  // Request interceptor: attach token + tenant schema
  api.interceptors.request.use((req) => {
    if (typeof window !== 'undefined') {
      const token = localStorage.getItem(AppConstants.storageKeys.accessToken)
      if (token) {
        req.headers.Authorization = `Bearer ${token}`
      }
      if (!isPublicEndpoint(req.url || '')) {
        const schema = localStorage.getItem(AppConstants.storageKeys.tenantSchema)
        if (schema) {
          req.headers['X-Tenant-Schema'] = schema
        }
      }
    }
    return req
  })

  // Response interceptor: pass errors through (matches Flutter behaviour)
  api.interceptors.response.use(
    (res) => res,
    (err) => Promise.reject(err)
  )

  return {
    provide: {
      api
    }
  }
})
